src/features/otsu_split.rs

```diff
@@ -0,0 +1,159 @@
+use crate::evaluator::*;
+use crate::time_series::DataSample;
+use conv::prelude::*;
+
+macro_const! {
+    const DOC: &'static str = r#"
+Otsu threshholding algorithm
+
+Difference of subset means, standard deviation of the lower subset, standard deviation of the upper
+subset and lower-to-all observation count ratio for two subsets of magnitudes obtained by Otsu's
+method split. Otsu's method is used to perform automatic thresholding. The algorithm returns a
+single threshold that separate values into two classes. This threshold is determined by minimizing
+intra-class intensity variance, or equivalently, by maximizing inter-class variance.
+
+- Depends on: **magnitude**
+- Minimum number of observations: **2**
+- Number of features: **4**
+
+Otsu, Nobuyuki 1979. [DOI:10.1109/tsmc.1979.4310076](https://doi.org/10.1109/tsmc.1979.4310076)
+
+Matwey Kornilov Otsu thresholding algorithm realization was used as a reference:
+    http://curl.sai.msu.ru/hg/home/matwey/domecam/file/tip/include/otsu.h
+    https://ieeexplore.ieee.org/document/9170791
+"#;
+}
+
+#[doc = DOC ! ()]
+#[derive(Clone, Default, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct OtsuSplit {}
+
+lazy_info!(
+    OTSU_SPLIT_INFO,
+    OtsuSplit,
+    size: 4,
+    min_ts_length: 2,
+    t_required: false,
+    m_required: true,
+    w_required: false,
+    sorting_required: false,
+);
+
+impl OtsuSplit {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    pub fn doc() -> &'static str {
+        DOC
+    }
+}
+
+impl FeatureNamesDescriptionsTrait for OtsuSplit {
+    fn get_names(&self) -> Vec<&str> {
+        vec![
+            "otsu_mean_diff",
+            "otsu_std_lower",
+            "otsu_std_upper",
+            "otsu_lower_to_all_ratio",
+        ]
+    }
+
+    fn get_descriptions(&self) -> Vec<&str> {
+        vec!["difference between mean values of Otsu split subsets",
+             "standard deviation for observations below the threshold given by Otsu method",
+             "standard deviation for observations above the threshold given by Otsu method",
+             "ratio of quantity of observations bellow the threshold given by Otsu method to quantity of all observations"]
+    }
+}
+
+impl<T> FeatureEvaluator<T> for OtsuSplit
+where
+    T: Float,
+{
+    fn eval(&self, ts: &mut TimeSeries<T>) -> Result<Vec<T>, EvaluatorError> {
+        self.check_ts_length(ts)?;
+
+        if ts.m.get_max() == ts.m.get_min() {
+            return Err(EvaluatorError::FlatTimeSeries);
+        }
+
+        let mut dm = T::zero();
+        let mut w: usize = 0;
+        let mean = ts.m.get_mean();
+        let count = ts.lenu();
+        let msorted = ts.m.get_sorted();
+        let mut last_variance = T::zero();
+
+        for &m in msorted.iter() {
```

Comment:
We can use `.enumerate()` or `(1..).zip(msorted.iter())` instead, in this case we don't need a mutable `w`